### ISSUE TITLE
Do not disclose links between contracts and hosts

### ISSUE THREAD

**Issue by starius** [NebulousLabs/Sia#2327]
Currently everybody who knows the contract id can determine the host which holds the files (see below how). This information can be used against the renter and the host. Contract ID (actually, the whole contract behind this ID) is a part of transactions and can be linked with renter's siacoins which can be linked to the renter identity in some cases. Also the contract can be linked to the host which can be linked to the location of the data. In other words with some luck the identity of the renter can be linked to the location of his or her data.

Another problem is that now everybody knows how much money the host makes from its contracts. Many people (including me) do not like when other people look into their wallets. There is also a problem of reusing the same address multiple times, which [is considered harmful](https://en.bitcoin.it/wiki/Address_reuse).

Currently there are 3 ways to link a contract to its host.

1. Some RPCs take a contract id and return different errors depending of whether they have the contract with this ID or not. Fixed in [PR 2304](https://github.com/NebulousLabs/Sia/pull/2304) (in review).

2. [FileContract.ValidProofOutputs and FileContract.ValidProofOutputs.MissedProofOutputs](https://godoc.org/github.com/NebulousLabs/Sia/types#FileContract) have host unlockhash. A renter takes it directly from the host by making RPCSettings and [uses it](https://github.com/NebulousLabs/Sia/blob/18bfef5e4b4e27b29603963e85d51f42f6623584/modules/renter/proto/formcontract.go#L77) when it creates the file contract transaction. By the way, in addition to possibility of linking contracts to hosts this results in reusing the same unlockhash in multiple outputs which [is considered harmful](https://en.bitcoin.it/wiki/Address_reuse). Currently hosts [return](https://github.com/NebulousLabs/Sia/blob/b3ef0cc99d070ba4ce3aa357f793c64b59d01336/modules/host/negotiatesettings.go#L44) the same unlockhash every time. It can instead generate a new one for each FormContract request and the same one for all RPCSettings. Possible abuse: somebody can send tons of FormContract requests without accepting the contract. It can result in tons of unused addresses in the output of `siac wallet addresses` of the host. This can be mitigated by putting unused unlockhashes back to the pool and reusing them. Fixes required: on the **host side only**.

3. [FileContract.UnlockHash](https://godoc.org/github.com/NebulousLabs/Sia/types#FileContract) is a hash of UnlockConditions [involving](https://github.com/NebulousLabs/Sia/blob/18bfef5e4b4e27b29603963e85d51f42f6623584/modules/renter/proto/formcontract.go#L31) the host PublicKey. The UnlockHash itself is not enough to infer the host (because it is a hash), but when a contract is revised (e.g. the renter uploads data), the public key of the host is disclosed as a part of [FileContractRevision.UnlockConditions](https://godoc.org/github.com/NebulousLabs/Sia/types#FileContractRevision). Currently renters take host public key from the [HostAnnouncement](https://godoc.org/github.com/NebulousLabs/Sia/modules#HostAnnouncement) struct which is announced to the blockchain by hosts. I think that instead the public key can be received from the host when a contract is being formed. This requires changes on **both renter and host sides** and probably new RPC.

**Comment by lukechampine** [comment]
I agree that we should do what we can to make such information private, but keep in mind that Sia was never intended to be a privacy-first blockchain. If privacy features significantly degrade the performance of the network, or if they make it much more difficult to become a host or renter, then we may reject them or make them opt-in.

1 and 2 definitely seem doable. 3 is a little more complicated. The pubkey seen in the host announcement needs to be used to bootstrap any new pubkeys. I'll wait for @DavidVorick to weigh in on this one.

**Comment by starius** [comment]
> The pubkey seen in the host announcement needs to be used to bootstrap any new pubkeys.

A contract is not tied to the announcement on the blockchain level, if I understand correctly. So the host can make a new pubkey controlling only this particular contract and send a signature of this key to the renter (with `crypto.WriteSignedObject`). The renter can verify that the new pubkey belongs to the same host using the pubkey from the announcement, but the renter has no incentive to disclose this link to 3d parties.

**Comment by tbenz9** [comment]
Recommendation: tag as 'Feature Request'.

**Comment by starius** [comment]
Recommendation: tag as 'Privacy'.

**Comment by DavidVorick** [comment]
Labeling this as 'Protocol Change', as we could not complete this without changing the protocol. 

At this point in Sia development, we don't have the resources or attention to manage protocol changes, though it's something we intend to do a lot more of later in 2018.
